feat!: upgrade to mathjax v4

navinkarkera opened the pull request:
### Description

Upgrades MathJax from **v2.7.5 → v4.1.2** across LMS, CMS, and xblock iframes.

Key changes:
- **API replacement** — `MathJax.Hub` calls replaced by `window.MathJax` config + `startup.promise` / `typesetPromise` / `typesetClear`
- **Output processor** — SVG → **CHTML** (Common HTML output, faster + CSS-driven)
- **Delimiter compatibility preserved** — `\(...\)`, `\[...\]`, `[mathjaxinline]...[/mathjaxinline]`, `[mathjax]...[/mathjax]`
- **AsciiMath** — CAPA preview backtick parsing unchanged

Co-authored by: GPT 5.5(xhigh)

### Supporting information

* `Private-ref`: [BB-7992](https://tasks.opencraft.com/browse/BB-7992)
* [MathJax v2 to v3 upgrade document](https://docs.mathjax.org/en/latest/upgrading/v2.html)
* xblocks-contrib: https://github.com/openedx/xblocks-core/pull/270

### Testing instructions

### Setup

```bash
cd /path/to/edx-platform
git checkout navin/upgrade-mathjax-v3
tutor mounts add /path/to/edx-platform
tutor images build openedx-dev && tutor dev launch -I --skip-build
tutor dev start -d
```

### 1. Verify MathJax v4 is loaded

Open browser DevTools → **Network tab**, filter for `mathjax`. Confirm **`mathjax@4.1.2/tex-mml-chtml.js`** is loaded in both **LMS** and **Studio**, and no v2 files.

### 2. Studio content editing


### HTML Block

Create a Raw HTML block with:

```xml
<html display_name="MathJax HTML Test">
  <p>Inline math: [mathjaxinline]E=mc^2[/mathjaxinline]</p>
  <p>Display math: [mathjax]\int_0^\infty e^{-x^2}\,dx = \frac{\sqrt{\pi}}{2}[/mathjax]</p>
  <p>Fraction: [mathjaxinline]3/4[/mathjaxinline]</p>
  <p>Quadratic: [mathjax]x = \frac{-b \pm \sqrt{b^2 - 4ac}}{2a}[/mathjax]</p>
</html>
```

### Problem Block — Text + Formula Input

Paste into Advanced Editor:

```xml
<problem display_name="MathJax Problem Test" markdown="null">
  <p>Simplify <strong>[mathjaxinline]3/4[/mathjaxinline]</strong>.</p>
  <p>The quadratic formula: <strong>[mathjax]x = \frac{-b \pm \sqrt{b^2 - 4ac}}{2a}[/mathjax]</strong></p>
  <p>What is <strong>[mathjaxinline]x^2[/mathjaxinline]</strong> when <strong>[mathjaxinline]x=3[/mathjaxinline]</strong>?</p>
  <numericalresponse answer="9">
    <formulaequationinput label="9"/>
  </numericalresponse>
</problem>
```

### Problem Block — Formula Preview Only

```xml
<problem display_name="MathJax Formula Preview" markdown="null">
  <p>Try <strong>x^2+1</strong>, <strong>x*y</strong>, <strong>1/2</strong>, <strong>(x+1)/(x-1)</strong>, <strong>sqrt(x)</strong>.</p>
  <formularesponse type="ci" samples="x@1:5" answer="x^2+1">
    <label>Enter expression:</label>
    <formulaequationinput size="40"/>
  </formularesponse>
  <solution>
    <div class="detailed-solution">
      <p>Explanation</p>
      <p>The expression <strong>[mathjaxinline]x^2+1[/mathjaxinline]</strong> is a quadratic.</p>
    </div>
  </solution>
</problem>
```

### Discussion Posts

Post #1 — Inline: `I was wondering about the famous equation $E=mc^2$.`

Post #2 — Block: `The quadratic formula: $$x = \frac{-b \pm \sqrt{b^2 - 4ac}}{2a}$$`

Post #3 — Mixed: `Trig: $\sin^2\theta + \cos^2\theta = 1$. Also: $\mu$ for mean, $\sigma$ for std dev.`

---

## Pass / Fail Criteria

- **Pass:** All math renders as formatted equations. No raw `[mathjax]`, `[mathjaxinline]`, `$`, or `$$` delimiters visible in student view.
- **Pass:** Formula equation input preview updates on every keystroke. Clear input → preview clears, no errors.
- **Pass:** Discussion posts render math inline and block correctly. Thread list gracefully handles math in preview text.
- **Pass:** Browser console shows **zero** errors matching:
  - `MathJax.Hub` is undefined / `Hub.Queue` is not a function
  - `MathJax.Hub.getAllJax` is not a function
  - `MathJax.Callback` is undefined
  - `typesetClear` / `typesetPromise` is not a function
  - `Cannot read properties of undefined (reading 'root')`
- **Fail:** Any of the above console errors, raw tags visible, or preview not updating.


### Notes / known follow-ups

- **v4 CHTML output is CSS-driven** — verify `.mjx-container` overflow and responsive width in all themes.
- **Companion PR** — xblocks-core#270 must be merged before for full non-iframe xblock coverage.
- **No intentional workflow changes** — visual output may differ slightly because v4 uses CHTML instead of SVG.
- **Performance** — v4 startup is faster; CHTML avoids SVG DOM overhead.
- **Iframe xblocks** — tested via the iframe rendering path; if an xblock bypasses the standard pipeline, check its own MathJax loading.


**Please let me know if we are using mathjax in some other components which we need to test**.

### Deadline

None

### Other information

None

navinkarkera committed:
feat: mathjax upgrade v2 to v3

fix: use asciimath

test: fix mathjax related tests

fix: lint issues

fix: mathjax_delay_renderer

test: fix mathjax related tests

openedx-webhooks commented:
Thanks for the pull request, @navinkarkera!


This repository is currently maintained by `@openedx/wg-maintenance-edx-platform`.


Once you've gone through the following steps feel free to tag them in a comment and let them know that your changes are ready for engineering review.

<details><summary>:radio_button: Get product approval</summary>

If you haven't already, [check this list](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Does-my-contribution-require-Product-Review%3F) to see if your contribution needs to go through the product review process.

- If it does, you'll need to submit a product proposal for your contribution, and have it reviewed by the [Product Working Group](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3449028609/Product+Working+Group).
    - This process (including the steps you'll need to take) is documented [here](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Product-Review-Process).
- If it doesn't, simply proceed with the next step.
</details>

<details><summary>:radio_button: Provide context</summary>

To help your reviewers and other members of the community understand the purpose and larger context of your changes, feel free to add as much of the following information to the PR description as you can:

- Dependencies
  > This PR must be merged before / after / at the same time as ...
- Blockers
  > This PR is waiting for OEP-1234 to be accepted.
- Timeline information
  > This PR must be merged by XX date **because** ...
- Partner information
  > This is for a course on edx.org.
- Supporting documentation
- Relevant [Open edX discussion forum](https://discuss.openedx.org/) threads
</details>





<details><summary>:radio_button: Get a green build</summary>

If one or more checks are failing, continue working on your changes until this is no longer the case and your build turns green.
</details><details>

<summary>:radio_button: Update the status of your PR</summary>

Your PR is currently marked as a draft.  After completing the steps above, update its status by clicking "Ready for Review", or removing "WIP" from the title, as appropriate.
</details>
<!-- comment:end_of_wip -->


---

<details><summary>Where can I find more information?</summary>

If you'd like to get more details on all aspects of the review process for open source pull requests (OSPRs), check out the following resources:

- [Overview of Review Process for Community Contributions](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/FAQ-about-pull-requests.html)
- [Pull Request Status Guide](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/pull-request-statuses.html)
- [Making changes to your pull request](https://docs.openedx.org/en/latest/documentors/how-tos/make_changes_to_your_pull_request.html)
</details>

<details><summary>When can I expect my changes to be merged?</summary>

Our goal is to get community contributions seen and reviewed as efficiently as possible.

However, the amount of time that it takes to review and merge a PR can vary significantly based on factors such as:

- The size and impact of the changes that it introduces
- The need for product review
- Maintenance status of the parent repository

:bulb: *As a result it may take up to several weeks or months to complete a review and merge your PR.*
</details>
<!-- comment:external_pr -->
<!-- data: eyJkcmFmdCI6IHRydWV9 -->

CefBoud commented:
@navinkarkera Great work on this!
Mathjax2 is also used in [frontend-app-library-authoring](https://github.com/openedx/frontend-app-library-authoring/blob/c75f8771b719e00e866cacd65ef85bc0c8f72eb6/src/library-authoring/edit-block/LibraryBlock/wrap.js#L362). Can you please take a look?

CefBoud commented on common/static/js/capa/src/formula_equation_preview.js:
```diff
@@ -169,30 +169,22 @@ formulaEquationPreview.enable = function() {
+                    // Set the text as the latex code, and then update the MathJax.
+                    MathJax.typesetClear([previewElement]);
+                    // Surround eqn with backticks "`", so that mathjax can process it as asciimath
+                    previewElement.innerHTML = `\`${latex}\``;
```
Is there a reason why this is processed as asciimath and not as latex as it was previously the case? 
Also, it is a bit confusing parsing a variable named `latex` as asciimath.

navinkarkera replied:
@CefBoud It was also the case before, AFAIK using back-ticks allow us to use both asciimath and tex commands. In version 3, the raw commands are stored in an external list of math objects while in version 2, it was stored in a `<script>` tag next to the rendered version so it was possible to update the raw command directly without removing elements.

> it is a bit confusing parsing a variable named latex as asciimath.

As mentioned above, it can also parse and render latex commands but `\[` & `\(` delimiters can only parse tex commands.

CefBoud replied:
> As mentioned above, it can also parse and render latex commands but \[ & \( delimiters can only parse tex commands.

@navinkarkera  Thank you for the explanation. I understand the reasoning.

Nit: however  if I am not mistaken, the input of the display function can only be latex because it goes through [preview_formcalc](https://github.com/openedx/edx-platform/blob/3bda3bed4ae2007763ce9197ba5258618f95d093/common/static/js/capa/src/formula_equation_preview.js#L125-L126) first which ends up calling [latex_preview ](https://github.com/openedx/edx-platform/blob/3bda3bed4ae2007763ce9197ba5258618f95d093/xmodule/capa/inputtypes.py#L1307), so I am not sure if there is a scenario in which the input of `display`  ends up being an asciimath expression.

navinkarkera replied:
@CefBoud You are right :100:, I just did not see any disadvantages with supporting both in frontend (maybe we add support for it in backend in future :smiley_cat:). Still, if you feel like we should change it to `\(`, I'll update it.

CefBoud replied:
@navinkarkera 
I think it would be best if it's `\(` to avoid any confusion. So I'd advise the change if you don't mind. 🙂

navinkarkera replied:
@CefBoud Done.

navinkarkera commented:
> Mathjax2 is also used in [frontend-app-library-authoring](https://github.com/openedx/frontend-app-library-authoring/blob/c75f8771b719e00e866cacd65ef85bc0c8f72eb6/src/library-authoring/edit-block/LibraryBlock/wrap.js#L362). Can you please take a look?

@CefBoud Although I don't know if it is in the scope, I can update it as it seems to be used in only once place.

CefBoud reviewed (approved):
:+1:

- [X] I tested this: in Text, Numerical Input, Math Expression and Problem components. Also in course Discussions. Inline and display expressions.
- [X] I read through the code
- [X] I checked for accessibility issues : Mathjax built-in accessibility functioning properly. Tested with Chrome Screen Reader extension.
- [ ] ~~Includes documentation~~
- [ ] ~~I made sure any change in configuration variables is reflected in the corresponding client's `configuration-secure` repository~~.

wittjeff commented:
Please see discussion re upgrade paths for MathML rendering not being simple in this earlier PR: https://github.com/openedx/edx-platform/pull/32418
If you'd like to implement a course-level Advanced Option (with at least MathJax v2.7.9 and 3.2.1), that'd be super.

navinkarkera commented:
@wittjeff Thanks for pointing me to the PR. About implementing course level advance option to support both v2 and v3, I'll come back with an update/answer soon.

wittjeff commented on cms/djangoapps/pipeline_js/js/xmodule.js:
```diff
@@ -16,41 +16,42 @@ define(
         $script(
-            'https://cdn.jsdelivr.net/npm/mathjax@2.7.5/MathJax.js'
-            + '?config=TeX-MML-AM_SVG&delayStartupUntil=configured',
+            'https://cdn.jsdelivr.net/npm/mathjax@3.2.1/es5/tex-mml-svg.js',
```
There is a v3.2.2 available. https://github.com/mathjax/MathJax/releases/tag/3.2.2

wittjeff commented on cms/static/cms/js/require-config.js:
```diff
@@ -137,7 +152,7 @@
 
             // externally hosted files
-            mathjax: 'https://cdn.jsdelivr.net/npm/mathjax@2.7.5/MathJax.js?config=TeX-MML-AM_SVG&delayStartupUntil=configured', // eslint-disable-line max-len
+            mathjax: 'https://cdn.jsdelivr.net/npm/mathjax@3.2.1/es5/tex-mml-svg.js?noext',
```
3.2.2 available

wittjeff commented on cms/static/cms/js/spec/main.js:
```diff
@@ -69,7 +100,7 @@
             'URI': 'xmodule_js/common_static/js/vendor/URI.min',
             'mock-ajax': 'xmodule_js/common_static/js/vendor/mock-ajax',
-            mathjax: 'https://cdn.jsdelivr.net/npm/mathjax@2.7.5/MathJax.js?config=TeX-MML-AM_SVG&delayStartupUntil=configured', // eslint-disable-line max-len
+            mathjax: 'https://cdn.jsdelivr.net/npm/mathjax@3.2.1/es5/tex-mml-svg.js?noext',
```
3.2.2 available

wittjeff commented on cms/static/cms/js/spec/main_squire.js:
```diff
@@ -48,7 +80,7 @@
             'domReady': 'xmodule_js/common_static/js/vendor/domReady',
             'URI': 'xmodule_js/common_static/js/vendor/URI.min',
-            mathjax: 'https://cdn.jsdelivr.net/npm/mathjax@2.7.5/MathJax.js?config=TeX-MML-AM_SVG&delayStartupUntil=configured', // eslint-disable-line max-len
+            mathjax: 'https://cdn.jsdelivr.net/npm/mathjax@3.2.1/es5/tex-mml-svg.js?noext',
```
3.2.2 available

wittjeff commented on cms/static/sass/studio-main-v1.scss:
```diff
@@ -18,3 +18,12 @@
 @import 'build-v1'; // shared app style assets/rendering
+
+
+.MathJax>svg {
```
I thought v3 eliminated SVG rendering. No?  SVG rendering is problematic when used as items in a select element.

wittjeff commented on cms/static/sass/studio-main-v1.scss:
```diff
@@ -18,3 +18,12 @@
+    max-width: 100%,
+}
+
+table>tbody>tr>td>.MathJax>svg {
```
Not sure if relevant now.

wittjeff commented on cms/templates/content_libraries/xblock_iframe.html:
```diff
@@ -113,48 +124,21 @@
        It can't be run through static.url because MathJax uses crazy url introspection to do lazy loading of
        MathJax extension libraries -->
-  <script type="text/javascript" src="https://cdn.jsdelivr.net/npm/mathjax@2.7.5/MathJax.js?config=TeX-MML-AM_SVG"></script>
+<script type="text/javascript" id="MathJax-script" src="https://cdn.jsdelivr.net/npm/mathjax@3.2.1/es5/tex-mml-svg.js"> </script>
```
3.2.2 available

wittjeff commented on common/templates/xblock_v2/xblock_iframe.html:
```diff
@@ -113,48 +124,21 @@
-          set_mathjax_display_div_settings();
-      }
-    });
-    function set_mathjax_display_div_settings() {
```
Are you sure this is all obsolete? I didn't see anything mentioned in the MathJax upgrade docs.  Is this custom code we added?

wittjeff commented on cms/templates/content_libraries/xblock_iframe.html:
```diff
@@ -113,48 +124,21 @@
-            })]
-          );
+          MathJax.typesetClear();
+          MathJax.typesetPromise().then(() => $('.MathJax>svg').toArray().forEach(el => {
```
SVG rendering??

wittjeff commented on common/static/common/js/discussion/mathjax_include.js:
```diff
@@ -1,56 +1,40 @@
+                ],
+                autoload: {
+                    color: [],
+                    colorv2: ['color']
```
Do we need to consider accessibility (particularly contrast) of a MathJax color palette?

wittjeff commented on common/static/common/js/discussion/mathjax_include.js:
```diff
@@ -1,56 +1,40 @@
-        }
-    };
-    vendorScript.src = 'https://cdn.jsdelivr.net/npm/mathjax@2.7.5/MathJax.js?config=TeX-MML-AM_HTMLorMML';
+    vendorScript.src = 'https://cdn.jsdelivr.net/npm/mathjax@3.2.1/es5/tex-mml-svg.js';
```
3.2.2 available

wittjeff commented on common/templates/mathjax_include.html:
```diff
@@ -131,5 +103,5 @@
      It can't be run through static.url because MathJax uses crazy url introspection to do lazy loading of
      MathJax extension libraries -->
-<script type="text/javascript" src="https://cdn.jsdelivr.net/npm/mathjax@2.7.5/MathJax.js?config=TeX-MML-AM_SVG"></script>
+<script type="text/javascript" id="MathJax-script" src="https://cdn.jsdelivr.net/npm/mathjax@3.2.1/es5/tex-mml-svg.js"></script>
```
3.2.2 available

wittjeff commented on lms/static/lms/js/spec/main.js:
```diff
@@ -56,7 +88,7 @@
             'jasmine-imagediff': 'xmodule_js/common_static/js/vendor/jasmine-imagediff',
             'domReady': 'xmodule_js/common_static/js/vendor/domReady',
-            mathjax: 'https://cdn.jsdelivr.net/npm/mathjax@2.7.5/MathJax.js?config=TeX-MML-AM_HTMLorMML&delayStartupUntil=configured', // eslint-disable-line max-len
+            mathjax: 'https://cdn.jsdelivr.net/npm/mathjax@3.2.1/es5/tex-mml-svg.js?noext',
```
3.2.2 available

wittjeff commented on lms/static/sass/lms-course.scss:
```diff
@@ -24,3 +24,11 @@
     }
 }
+
+.MathJax>svg {
```
Use of SVG??

wittjeff commented:
Related discussion re Canvas use of MathJax: https://community.canvaslms.com/t5/Canvas-Question-Forum/Canvas-s-MathJax-Needs-An-Update/m-p/583189

Yagnesh1998 commented:
I submitted a [PR](https://github.com/openedx/edx-platform/pull/32418) because I'm having some issues with MathJax on the edX platform.Let me know if there is anything I can do here

rayzhou-bit commented:
Hi @navinkarkera, I tried testing this recently with a more current version of edx-platform and could not get it to work. Is it still working on your environment?

navinkarkera commented:
@rayzhou-bit Thanks for testing it. I am currently occupied with another project and we haven't received any confirmation from the original stakeholders yet, but I hope to get back to this soon.

OmarIthawi reviewed (changes requested):
Thanks @navinkarkera for your contribution. This is really useful and the update is much needed to improve performance.

I recommend using HTML CSS instead of SVG whenever possible. SVG is challenging for [MathJax plugin makers](https://github.com/OmarIthawi/arabic-mathjax) like myself when it comes to Arabic and RTL support.

Additionally, Open edX needs to support mobile and to stay accessible which I think MathJax SVG isn't very well suited for:

 - https://docs.mathjax.org/en/latest/output/html.html

mphilbrick211 commented:
Hi @navinkarkera! Are you planning to pursue this PR?

navinkarkera commented:
@mphilbrick211 Yes, once mathjax 4 is released, we plan to upgrade since it is already in beta. For more info, please see this [discussion](https://discuss.openedx.org/t/mathjax-3-upgrade-interest/13387/3)

mphilbrick211 commented:
Hi @navinkarkera! Just checking in on this :)

navinkarkera commented:
@mphilbrick211 Sorry, I don't have any update here.

wittjeff commented:
Version 4.0.0 is [now available](https://github.com/mathjax/MathJax-src).

feanil commented:
@navinkarkera I'm gonna move this to draft until you're ready to with the MathJax 4.0 update.  Once it's ready for review let me know and I can help get it reviewed and landed.

navinkarkera committed:
Merge branch 'master' into navin/upgrade-mathjax-v3

navinkarkera committed:
fix(mathjax): complete v4 upgrade fixes

- Pin CDN URLs to mathjax@4.1.2
- Fix mathjax_delay_renderer.js immediate path to pass DOM node
- Add null guard in capa/display.js for missing math element
- Fix CSS syntax: comma to semicolon in MathJax SVG rules
- Pass textContent instead of innerHTML for security

navinkarkera committed:
fix(mathjax): use chtml output

navinkarkera committed:
fix(mathjax): support v4 formula previews

navinkarkera committed:
fix(mathjax): render formula previews with v4

- mathjax_delay_renderer.js: add typesetClear before .html() in both immediate and delayed paths, so MathJax v3/v4 re-renders replaced content
- capa/display.js refreshMath: replace v2 Hub.Queue/getAllJax with v3/v4 typesetClear/typesetPromise; add typesetClear before setting preview textContent; unify rendering to AsciiMath/backtick for all non-tex-delimited input (fixes formula preview for formulaequationinput where * → multiplication and / → fraction instead of literal chars)
- capa/display_spec.js: add tests for calculator syntax backtick wrapping, TeX delimiter preservation, and empty-input early return

navinkarkera committed:
Merge branch 'master' into navin/upgrade-mathjax-v3

navinkarkera committed:
temp: fix

navinkarkera committed:
fix(mathjax): enable extracted xblocks for v4

navinkarkera committed:
Merge branch 'master' into navin/upgrade-mathjax-v3

navinkarkera committed:
fix(mathjax): stabilize v4 xblock rendering

navinkarkera committed:
Merge branch 'master' into navin/upgrade-mathjax-v3

navinkarkera committed:
Merge branch 'master' into navin/upgrade-mathjax-v3

navinkarkera committed:
fix(mathjax): restore dollar delimiters

navinkarkera committed:
fix(mathjax): remove redundant dollar delimiter config and detection

Dollar delimiters ($...$, $$...$$) are MathJax 3 defaults, so explicit config is redundant. Remove them from config, inline detection, and tests, and simplify the tag-detection logic (no remaining self-matching pairs).

navinkarkera committed:
fix: remove invalid settings

navinkarkera committed:
Merge branch 'master' into navin/upgrade-mathjax-v3

navinkarkera committed:
Merge branch 'master' into navin/upgrade-mathjax-v3
